compose: Redesign and relocate compose expand / collapse button.

N-Shar-ma opened the pull request:
compose: Redesign and relocate compose expand / collapse button.

The compose expand / collapse button is moved from the top right of the compose area to the top right corner inside the compose textarea / preview.

The textarea / preview and the button area shares the same grid parent. 25px extra padding is added to and margin is subtracted from the right of the element so the button is visually inside it. It appears only on hovering the compose area, and uses a new custom icon.

Fixes: #28791.


### Screenshots and screen captures:

#### Dark:
**Collapsed state: (editing mode)**
(firefox on linux)
![image](https://github.com/zulip/zulip/assets/68962290/82963d31-3ee8-4f3b-a159-524a6fceb40a)
![image](https://github.com/zulip/zulip/assets/68962290/4fb78906-fe15-4048-a44e-9507084483be)
(chrome on linux)
![image](https://github.com/zulip/zulip/assets/68962290/838e75b9-7fd2-4f91-bdd3-44e3834c107f)
![image](https://github.com/zulip/zulip/assets/68962290/07d38306-6b9a-4983-829e-341c76766676)

**Expanded state: (preview mode)**
![image](https://github.com/zulip/zulip/assets/68962290/4ac02dd8-ebc5-4271-b854-7af19510c9f7)

#### Light:
**Collapsed state:** (editing mode in chrome on linux)
![image](https://github.com/zulip/zulip/assets/68962290/14070c61-588c-4607-a5f5-221b2e97a7f7)
![image](https://github.com/zulip/zulip/assets/68962290/4a86886c-e55a-4808-80aa-c8140928b0ad)

**Expanded state: (preview mode)**
![image](https://github.com/zulip/zulip/assets/68962290/4b8491fb-e938-416f-82f7-d83fcd245201)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

N-Shar-ma commented:
@alya This is ready for initial testing from you

alya commented:
Thanks! The "expand" button looks good, but something is off about the "collapse" button. One of the arrows is thicker than the other, and it's not centered within the hover area.

![CleanShot 2024-04-11 at 12 34 14@2x](https://github.com/zulip/zulip/assets/2090066/7c62f2a8-4a8e-426a-9b27-713c84cefd9d)
![CleanShot 2024-04-11 at 12 34 04@2x](https://github.com/zulip/zulip/assets/2090066/988ece85-d2aa-426d-90bd-8f0967caa0c6)

alya commented:
Also, while we're here, let's change the tooltips to "Expand compose box" and "Collapse compose box".

alya commented:
And let's include updating https://zulip.com/help/resize-the-compose-box as part of this PR.

N-Shar-ma commented:
@alya I'll work on and push the documentation changes tomorrow, but have fixed the icon centering and updated the tooltip text. As for the uneven line widths. I don't see it, maybe it's your secondary monitor? If not, @terpimost could you check the svg icon (taken from the [live demo](https://terpimost.github.io/compose-decomposed/))?

terpimost commented:
Hey, sorry for the icons (demo was created a long time ago). Here are corrected svgs:
[minimize](https://github.com/zulip/zulip/assets/1903309/a8059a87-68a3-4852-bd7e-d4264f2544bd)
[maximize](https://github.com/zulip/zulip/assets/1903309/ecdb024e-6c23-4fe0-91bd-11905d7ed32f)
[enlarge](https://github.com/zulip/zulip/assets/1903309/4de30436-8a82-4dda-aeab-fc0deadd2ab8)

I recorded some clarifications on a video https://cln.sh/KsBwrnjw

karlstolley commented on web/styles/compose.css:
```diff
@@ -235,6 +235,7 @@
         box-shadow: none;
         border: none;
         height: 100%;
+        position: relative;
```
That new element would then need to be positioned in the compose-box grid rather than the textarea, of course.

N-Shar-ma replied:
Thanks a lot for the suggestion! I had initially tried attempting it this way but couldn't get the height stretching in expanded mode to work. Took some time off this issue then worked on the restructuring as a separate issue, independent from the button relocation, and it was so much easier! :sweat_smile: 

The restructuring is in its own prep commit now.

N-Shar-ma commented on web/styles/compose.css:
```diff
@@ -266,8 +266,46 @@
+}
+
+#message-content-container .composebox-buttons {
+    z-index: 1;
```
This z index is needed as otherwise hovering when previewing causes weird flickering of the button and the cursor.

karlstolley replied:
Let's be sure to document that in the code here.

N-Shar-ma replied:
Done!

N-Shar-ma commented on help/resize-the-compose-box.md:
```diff
@@ -10,17 +10,18 @@ box also stretches automatically when you type a long message.
-1. Click on the **<i class="fa fa-chevron-up"></i> icon** in the top
-   right corner of the compose box.
+1. Click on the **<i class="zulip-icon zulip-icon-expand-diagonal"></i> icon** in the top
+   right corner of the compose box, which is visible on hovering over the message content area.
```
I looked through the help center articles but it doesn't seem we have a dedicated term for the textarea / preview, and we just call it the compose box, same as the overall area. Since it needed to be distinguished here to specify where exactly to hover, I call it "message content area"... @alya Let me know if something else would be better!

N-Shar-ma commented:
Finally addressed all the suggestions! Have updated the screenshots and commit messages, so this is ready for another round of testing and review.

@terpimost Thanks for the pointers, the buttons' styling and behavior should now be near identical to that in the demo. As for the 3 stage expansion, that I think is better as a follow up (thoughts @alya?)

karlstolley commented on web/styles/compose.css:
```diff
@@ -263,14 +263,51 @@
+    grid-area: message-content-container;
+    display: grid;
+    grid-template: minmax(0, 1fr) / minmax(0, 1fr) 25px;
+    margin-top: 5px;
```
I'd love to see this `margin-top` go away in some future refactor that uses grid in place of flex on `.compose_table`. But for the time being, we can actually get a better vertically-aligned Drafts button if that `margin-top` is moved over to `.messagebox`:

![Screenshot 2024-04-24 at 2 34 27 PM](https://github.com/zulip/zulip/assets/170719/f9ea9e8e-ddab-49c8-9cd0-d8a519172721)

karlstolley commented on web/templates/compose.hbs:
```diff
@@ -76,14 +74,20 @@
+                                        <div class="preview_content rendered_markdown"></div>
+                                    </div>
                                 </div>
+                                <div class="composebox-buttons">
```
So question: will you be able to reuse this `.compose-box` DIV for the formatting buttons? If not, it might be better to choose a more specific class name here--though I guess `composebox-formatting-buttons` is the likely candidate for those. Just something to think about, especially given the grid that you're setting up on this.

N-Shar-ma replied:
Yeah, I am not really able to come up with a better name for these... My thinking here was that these are compose *box* buttons while the formatting ones are compose *control* buttons.

Do you have any better suggestions in mind?

timabbott replied:
Maybe `composebox-textarea-side-buttons` would be good.

karlstolley commented on web/styles/compose.css:
```diff
@@ -704,16 +740,16 @@
     color: hsl(0deg 0% 33%);
     background-color: hsl(0deg 0% 100%);
+    margin-right: -25px;
+    padding-right: 25px;
```
This doesn't feel very grid-like. Is there a reason to do this hackery with padding and a negative margin instead use of named grid-areas to position things where they belong?

N-Shar-ma replied:
The design needs these buttons to show up visually inside the textarea, that's why it does need the textarea to be extended using negative margin over to be the backdrop for the buttons.

An alternative could be to have the buttons' grid area be the same background color and same height, with the borders adjusted to look like it's the "same" textarea extended. This is what I have in mind for making the formatting buttons row seem to be "inside" the textarea as well. But here. the main problem is that _the scrollbar will appear weirdly between the textarea instead of being pushed to the right edge_. Hence I went with the negative margin approach here.

Does that make sense @karlstolley?

karlstolley replied:
Ah, got it. Looking at this more closely, it seems like the issue is also to keep text in the textarea from running behind the buttons? That's probably okay...I'd just ask that you document that next to the lines above so those values have some adjacent explanation.

N-Shar-ma replied:
Yep! Forgot to mention the padding keeps the text from overlapping with the buttons' area. Have added a comment explaining all this

karlstolley replied:
Got it. And it occurs to me that, because that 25px value shows up in the grid template, too, it might be worth making a variable of those values so it's a) set in one place, and b) discoverable across the code base.

N-Shar-ma replied:
That makes perfect sense! I've defined a variable for this and moved the explanatory comment there too

karlstolley commented on web/styles/compose.css:
```diff
@@ -263,14 +263,51 @@
+#message-content-container {
+    grid-area: message-content-container;
+    display: grid;
+    grid-template: minmax(0, 1fr) / minmax(0, 1fr) 25px;
```
I'd really like to see named grid areas here and the textarea and buttons div placed on them explicitly. I think that will end up being important for when you go to get the formatting buttons in place, as those will likely need to take up the space in the column you've set up for the expand/collapse buttons. Plus it's just good practice in an area like this to use named areas for the sake of contributors who will have no idea how the expand/collapse buttons end up where they do.

karlstolley commented on web/styles/app_variables.css:
```diff
@@ -261,6 +261,7 @@
         231deg 100% 90% / 50%
     );
     --color-compose-chevron-arrow: hsl(0deg 0% 58%);
+    --color-composebox-button-background: hsl(0deg 0% 0% / 5%);
```
Let's add `-hover` to the end of these button-background variables, so that a) it's clear those are shown on hover and b) to future-proof against a design that has a background color outside the hover state (heck, it might be worth doing that now, even if you use the `transparent` value).

N-Shar-ma replied:
I myself was on the fence about this same decision... Have added a variable for transparent background too now!

karlstolley commented:
Left some feedback mostly concerning CSS Grid and things. They're largely nits, but nits worth taking care of before this gets merged into main.

N-Shar-ma commented:
@karlstolley Any next steps here?

timabbott commented:
This seems to merge conflict with the chat.zulip.org branch, so deferring for now.

timabbott commented:
Test-deployed on chat.zulip.org.

N-Shar-ma commented:
@timabbott While we test this on CZO, could you merge the 1st commit (no visual effect, just html restructuring)? That'd help me in working on other redesigns in the compose area without getting blocked on this PR.

timabbott commented:
Merged the first commit as 1121d317c26b0379f0131c3a246579d275e4a1d8.

alya commented:
I left some comments [on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20compose.20box.20expand.2Fcollapse.20button/near/1814240).

alya commented:
@timabbott My feedback has been addressed.

timabbott commented:
Merged, thanks @N-Shar-ma!
